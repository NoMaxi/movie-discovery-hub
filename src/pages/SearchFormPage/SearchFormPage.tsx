import { SearchForm } from "@/components/SearchForm/SearchForm";
import { Outlet, useSearchParams } from "react-router-dom";

export const SearchFormPage = () => {
    const [searchParams, setSearchParams] = useSearchParams();

    const searchQuery = searchParams.get("query") ?? "";

    const handleSearch = (query: string) => {
        const newParams = new URLSearchParams(searchParams);

        if (query) {
            newParams.set("query", query);
        } else {
            newParams.delete("query");
        }

        setSearchParams(newParams, { replace: true });
    };

    return (
        <div className="flex flex-col items-start px-[60px]">
            <h2 className="text-[40px] font-light uppercase mb-8 mt-[25px] tracking-[1px]">Find Your Movie</h2>
            <div className="w-full">
                <SearchForm initialQuery={searchQuery} onSearch={handleSearch} />
                <Outlet />
            </div>
        </div>
    );
};
