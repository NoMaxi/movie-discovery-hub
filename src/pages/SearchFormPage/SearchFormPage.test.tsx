import { render, screen } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { SearchFormPage } from "./SearchFormPage";

jest.mock("@/components/SearchForm/SearchForm", () => ({
    SearchForm: ({ initialQuery, onSearch }: { initialQuery: string; onSearch: (query: string) => void }) => (
        <div data-testid="search-form">
            <p data-testid="initial-query">{initialQuery}</p>
            <button type="button" onClick={() => onSearch("new-query")}>
                Search
            </button>
            <button type="button" onClick={() => onSearch("")}>
                Empty Search
            </button>
        </div>
    ),
}));

const mockSetSearchParams = jest.fn();

let mockedQueryString = "";
jest.mock("react-router-dom", () => {
    const actual = jest.requireActual("react-router-dom");
    return {
        ...actual,
        useSearchParams: jest.fn(() => [new URLSearchParams(mockedQueryString), mockSetSearchParams]),
    };
});

describe("SearchFormPage", () => {
    let user: UserEvent;

    beforeEach(() => {
        mockSetSearchParams.mockClear();
        mockedQueryString = "";
        user = userEvent.setup();
    });

    it("Should render correctly", () => {
        const { asFragment } = render(<SearchFormPage />);

        expect(screen.getByRole("heading", { level: 2, name: /find your movie/i })).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it("Should pass query from URL as 'initialQuery'", () => {
        mockedQueryString = "?query=Matrix";
        render(<SearchFormPage />);

        expect(screen.getByTestId("initial-query")).toHaveTextContent("Matrix");
    });

    it("Should update search params when a new search is submitted", async () => {
        render(<SearchFormPage />);

        await user.click(screen.getByRole("button", { name: "Search" }));

        expect(mockSetSearchParams).toHaveBeenCalledTimes(1);

        const [newParams, options] = mockSetSearchParams.mock.calls[0];

        expect(typeof newParams).toBe("object");
        expect(options).toEqual({ replace: true });
        expect(newParams.get("query")).toBe("new-query");
    });

    it("Should remove query parameter when empty search is submitted", async () => {
        mockedQueryString = "?query=Matrix";
        render(<SearchFormPage />);

        await user.click(screen.getByRole("button", { name: "Empty Search" }));

        expect(mockSetSearchParams).toHaveBeenCalledTimes(1);

        const [newParams, options] = mockSetSearchParams.mock.calls[0];

        expect(typeof newParams).toBe("object");
        expect(options).toEqual({ replace: true });
        expect(newParams.has("query")).toBe(false);
    });
});
