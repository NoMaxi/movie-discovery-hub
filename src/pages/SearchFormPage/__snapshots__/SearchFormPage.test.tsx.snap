// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFormPage Should render correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-start px-3xl"
  >
    <h2
      class="text-2xl font-light uppercase mb-8 mt-[25px] tracking-[1px]"
    >
      Find Your Movie
    </h2>
    <div
      class="w-full"
    >
      <div
        data-testid="search-form"
      >
        <p
          data-testid="initial-query"
        />
        <button
          type="button"
        >
          Search
        </button>
        <button
          type="button"
        >
          Empty Search
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
